app/src/main/java/com/android/streetworkapp/model/image/ImageRepositoryFirestore.kt

```diff
@@ -0,0 +1,115 @@
+) : ImageRepository {
+  companion object {
+    const val DEBUG_PREFIX = "ImageRepositoryFirestore:"
+    const val COLLECTION_PATH = "parkImages"
+  }
+
+  /**
+   * Uploads an image in b64 format into our database
+   *
+   * @param imageB64 The base64 encoded image
+   * @param parkId The parkId the image will be linked to
+   * @param userId The id of the image uploader
+   */
+  override suspend fun uploadImage(imageB64: String, parkId: String, userId: String) {
+    require(imageB64.isNotEmpty()) { "imageB64 should not be empty." }
+    require(parkId.isNotEmpty()) { "parkId cannot be empty." }
+    require(userId.isNotEmpty() && this.userRepository.getUserByUid(userId) != null) {
+      "Invalid userId."
+    }
+
+    try {
+      // checking if the park already has an image collection
+      val park =
+          this.parkRepository.getParkByPid(parkId)
+              ?: throw IllegalArgumentException("Invalid parkId")
+      val parkImage = ParkImageDatabase(imageB64 = imageB64, userId = userId)
+
+      val collectionId = park.imagesCollectionId
+      if (collectionId.isEmpty()) { // no imagesCollection for the park yet
+        val collectionRef = db.collection(COLLECTION_PATH).document()
+        val newCollectionId = collectionRef.id
+
+        val imagesCollection = ParkImageCollection(id = newCollectionId, listOf(parkImage))
+        // could have some loose collections if the call fails here, can't be bothered to handle
+        // this case
+        db.collection(COLLECTION_PATH).document(collectionRef.id).set(imagesCollection).await()
+        this.parkRepository.addImagesCollection(park.pid, newCollectionId)
+      } else { // there already exists an imagesCollection
+        val imagesCollection = db.collection(COLLECTION_PATH).document(collectionId)
+        val newImage = ParkImageDatabase(imageB64, userId)
+        imagesCollection.update("images", FieldValue.arrayUnion(newImage)).await()
+      }
+    } catch (e: Exception) {
+      Log.d(
+          DEBUG_PREFIX,
+          e.message
+              ?: "An exception occurred but the message associated with couldn't be retrieved.")
+    }
+  }
+
+  /**
+   * Retrieves all the images linked to the park.
+   *
+   * @param park The park linked to the images
+   */
+  override suspend fun retrieveImages(park: Park): List<ParkImageDatabase> {
+    require(this.parkRepository.getParkByPid(park.pid) != null) { "Invalid ParkId." }
+
+    if (park.imagesCollectionId.isEmpty())
+        return emptyList() // no collection setup yet, just return no images
+
+    try {
+      // checking the collection linked to the park is valid
+      val document = db.collection(COLLECTION_PATH).document(park.imagesCollectionId).get().await()
+      if (!document.exists())
+          throw IllegalArgumentException(
+              "The ImagesCollection linked by this park doesn't exist, the entry in the db is corrupted.")
+
+      val parkImages = document.get("images") as? List<*>
```

Comment:
Refactor: this "get" could be changed to a [] (similar to what was used in line 91 for instance)

Reply:
done